src/pairingheap.c

```diff
@@ -0,0 +1,114 @@
+/*
+ * Datastructures: GAP package providing common datastructures.
+ * Licensed under the GPL 2 or later.
+ *
+ * Implementation of a binary heap.
+ *
+ * Binary heaps are of course pretty standard data structures.
+ * However, a few design choices are possible. The implementation
+ * below has been influenced by this StackOverflow answer:
+ *   <https://stackoverflow.com/questions/6531543>
+ */
+
+#include "src/compiled.h"
+#include "pairingheap.h"
+
+#define DS_BINARYHEAP_ISLESS(heap) ELM_PLIST(heap, 1)
+#define DS_BINARYHEAP_DATA(heap) ELM_PLIST(heap, 2)
+
+
+static Obj SumSmallInt(Obj a, Obj b)
+{
+    GAP_ASSERT(IS_INTOBJ(a));
+    GAP_ASSERT(IS_INTOBJ(b));
+    Obj sum;
+    if (!SUM_INTOBJS(sum, a, b))
+        ErrorMayQuit("PANIC: counter overflow in pairing heap code", 0, 0);
+    return sum;
+}
+
+Obj DS_merge_pairs(Obj self, Obj isLess, Obj heaps)
+{
+    if (!IS_DENSE_PLIST(heaps))
+        ErrorQuit("<heaps> is not a dense plist", 0L, 0L);
+
+    const UInt len = LEN_PLIST(heaps);
+    Obj res;
+
+    if (len == 0) {
+        res = NEW_PLIST(T_PLIST_CYC_SSORT, 3);
```

Comment:
SSORT looks wrong here, since you initialise the list with 3 0's. Just T_PLIST_CYC?